### PR TITLE
fix(ios): multitap consistency after new-lines

### LINKED ISSUE #10696: bug(ios): potential context desync after multitaps + paragraph breaks
Possibly related to #10204 / #10633 and the related underlying behaviors.

Being able to elaborate further may be tricky without further investigation, but I'm pretty sure we can all agree that the following behavior, after the newline, is _wrong_.

This is an issue against `master` that is merely perpetuated and/or changed, but not resolved, by #10589 at present.

repro

Motivated by https://github.com/keymanapp/keyman/pull/10589#issuecomment-1940600108:
> Finally, I have identified what caused the issue in the PR. Here is what I did while following the steps:
> [...]
> 2. Next, I pressed the key to move to the next line.

Debugger-based inspection of the keyboard WebView indicates that, at the following frame:

image

... the KMW-internal context is:

Screen Shot 2024-02-13 at 3 21 19 PM

... or similar.  It appears that the attempt to resynchronize context for backspaces (see #10633) may be having adverse consequences as complicated new text is typed.  The `textDocumentProxy` context window also seems to be acting pretty oddly here, but that's just an inference at this time.

Other patterns sometimes results in all preceding context (having been maintained within the Web engine) suddenly being _duplicated_ - something that is _definitely_ concerning and erroneous behavior.

----

Repro:

1. Download this KMP: https://jahorton.github.io/diacritic_rota.kmp
    - Can also be found via the main page at https://jahorton.github.io.
2. Open / install it for use within the Keyman app.
3. Double-tap the `8` key.
    - Expected total result: `a`
4. Triple-tap the '◌̀' key - a circumflex diacritic should appear over the `a`.
5. Triple-tap the `9` key.
    - Expected total result: `aê`
6. Tap the '◌̀' key.  A diacritic should appear over the `ê`.
7. Open the in-app Settings menu.
    - Note, for test maintenance:  the point of this step and the next one is to trigger a context-reset.
8. Close the in-app Settings menu.
9. Double-tap the `8` key.
    - Expected final result:  `aềa`
    - For clarity:
        - With a copy of the `à`'s accent-mark on top the original `ê`.  
            - It looks different in raw-text form here due to GitHub's font / rendering.
10. Double-tap the ENTER key.
11. Repeat steps 3-6, then step 9.

Done using Simulator targeting iPhone 13 Pro, iOS 15.4.  Interestingly, it doesn't appear to repro in Simulator targeting iPhone SE (3rd gen), iOS 16.2?  At least... not the same sequence.  That said, it _was_ with an iPhone 13 Pro (non-simulated) where we saw weirdness in the tests for #10589, which led to discovery of this issue.

### PR DESCRIPTION
**Pull request by jahorton**
Fixes #10696.

Continues from #10584 and #10633.

Now that we're synchronizing app context with the internal web engine's context more proactively - which gives us more consistent backspace behavior - we need to ensure that this doesn't break preservation of context _going forward_, as _new_ text is added.

### Commit 1

_Simplified_:
- Ensures nothing "goes wrong" when you type a lot of text, even when starting new paragraphs. 
   - The lone exception: multitaps at the start of a new paragraph (and rarely, within the middle of a long one) may have unexpected side effects.

_Internal details_:
- Fixes context synchronization for all cases _except_ a multitap immediately following a `textDocumentProxy` context-window slide.

- Ensures that all context manipulations for each tap of a multitap requires only a single call when updating host-app context.  
    - Before, each tap of a multitap triggered two calls:  one to restore the old context, then one to apply the new tap's key.

### Commit 2

_Simplified_
- Fixes the paragraph-initial multitap issues mentioned for "commit 1".
- Prevents the engine from forgetting deadkeys too quickly when new paragraphs are started (which was an unexpected side effect of recent blocked-backspace fixes - #10633).

_Internal details_:
- Ensures `textDocumentProxy` context window jumps that result as new text is added (without other changes) do _not_ trigger a context reset within the internal web engine. 
    - Leverages #10662 to accomplish this goal.  I did _not_ expect to find another use for the method so soon, but here we are.
    - It is imperative that a context window jump does _not_ change the internal engine's context during an ongoing multitap.
    - Toward the former point:  we were getting excess newlines for context-window-initial multitaps - one per tap of the ongoing multitap gesture - because the context window was being reset between each tap due to `textDocumentProxy` quirks.
- This should also prevent context-resets during backspaces in which the `textDocumentProxy` context window returns to text that was once "in window" but had since slid out of it.

There is _a chance_ that an extreme edge case might still have issues:
- Suppose a keyboard rule that changes the last character before the newline preceding the caret.  (Say, `a\n` + `b` => `b\n`.)
- Now suppose that rule is triggered by a multitap key, thus cycling in and out during multitap rota alongside context rewind operations.
- As the tap with the rule would trigger an edit (within the host app) past context-window boundaries, this would naturally trigger context-synchronization after the related tap should the `textDocumentProxy` perform a context-window jump as a result... and also for the tap after it, which would need to make counter-edits to revert such a rule's change, as such a tap would likely retrigger a `textDocumentProxy` update.

Upon request, I suppose I could create a highly-focused test keyboard for validating such a case.  That said, it _does_ seem _pretty_ niche - I don't think we currently have a keyboard with such rules.  (As a result, I haven't prioritized testing this scenario.)  Part of me wonders if the best course of action would be to spin this concern off into a separate issue so that it doesn't hold this PR up from fixing a lot of cases that are far more commonly seen.

## User Testing

**TEST_LONG_DELETION**:  Using the iOS Keyman app on an iPhone 13 / iPhone 13 Pro, delete a _lot_ of pre-existing text from within a third-party application.  (Simulator is fine, but stick to iPhone 13 targets.)

1. Ensure that Keyman is enabled for use as a system keyboard.
2. Open a third party application **and swap to the default iOS system keyboard**.
3. Visit this PR's page on the test device and copy the full text from the code box above.  ("`Testing testing`...")
4. Open the Notes app and create a new note.
5. Paste the text copied from this PR's page as the content of the new note.
6. Use the globe key keyboard-selection menu to **activate Keyman as your system keyboard**.
7. Hold down BKSP until no further deletions occur.
8. If any visible text remains that refuses to be deleted, FAIL this test.

This test can be done on Simulator by using the Messages app instead and selecting one of the sample contacts.  The Notes app is typically not available for Simulator-based iOS devices.  (Notes doesn't come with the risk of butt-texting, though.)

**TEST_SHIFTED_CONTEXT_MULTITAP_IOS**: Using the iOS Keyman app on an iPhone 13 / iPhone 13 Pro, verify that multitaps at the start of new paragraphs do not have unwanted side effects.  (Simulator is fine, but stick to iPhone 13 targets.)

1. Download this KMP: https://jahorton.github.io/diacritic_rota.kmp
    - Can also be found via the main page at https://jahorton.github.io.
2. Open / install it for use within the Keyman app.
3. Double-tap the `8` key.
    - Expected total result: `a`
4. Triple-tap the '◌̀' key - a circumflex diacritic should appear over the `a`.
5. Triple-tap the `9` key.
    - Expected total result: `aê`
6. Tap the '◌̀' key.  A diacritic should appear over the `ê`.
7. Open the in-app Settings menu.
    - Note, for test maintenance:  the point of this step and the next one is to trigger a context-reset.
8. Close the in-app Settings menu.
9. Double-tap the `8` key.
    - Expected final result:  `aềa`
    - For clarity:
        - With a copy of the `à`'s accent-mark on top the original `ê`.  
            - It looks different in raw-text form here due to GitHub's font / rendering.
10. Double-tap the ENTER key.
11. Repeat steps 3-9.  The result for the first and last lines should be identical, with no extra lines produced after step 10.

Note:  this PR does not include the fix from #10589; things will break here if a diacritic is the first entry on a line.  This is known and expected.

**TEST_SHIFTED_CONTEXT_MULTITAP_ANDROID**: Using the Keyman for Android app, verify that multitaps at the start of new paragraphs do not have unwanted side effects.

1. Download this KMP: https://jahorton.github.io/diacritic_rota.kmp
    - Can also be found via the main page at https://jahorton.github.io.
2. Open / install it for use within the Keyman app.
3. Double-tap the `8` key.
    - Expected total result: `a`
4. Triple-tap the '◌̀' key - a circumflex diacritic should appear over the `a`.
5. Triple-tap the `9` key.
    - Expected total result: `aê`
6. Tap the '◌̀' key.  A diacritic should appear over the `ê`.
7. Open the in-app Settings menu.
    - Note, for test maintenance:  the point of this step and the next one is to trigger a context-reset.
8. Close the in-app Settings menu.
9. Double-tap the `8` key.
    - Expected final result:  `aềa`
    - For clarity:
        - With a copy of the `à`'s accent-mark on top the original `ê`.  
            - It looks different in raw-text form here due to GitHub's font / rendering.
10. Double-tap the ENTER key.
11. Repeat steps 3-9.  The result for the first and last lines should be identical, with no extra lines produced after step 10.

**TEST_NEW_PARAGRAPH_PREDICTIONS**: Using the iOS Keyman app on an iPhone 13 / iPhone 13 Pro, verify that predictive text works properly at the start of new paragraphs.

- Type a few paragraphs, making sure to use predictive text to apply a suggestion at the start of each.
- Type a few letters for the first word for some, but for others, immediately use a suggestion before typing anything (if possible).
- On occasion, add extra new lines between new paragraphs.
- If no suggestions are available after double-tapping ENTER, do not fail this test as a result.
   - It _might_ be worth an issue, though... and I do see this occurring in my local tests.
- If anything else unexpected occurs, or feels wrong, FAIL this test and take note of the issues.

Note:  this PR does not include the fix from #10589; things will break here if a diacritic is the first entry on a line.  This is known and expected.

### CODE REVIEW

**Comment by bharanidharanj** [comment]
Test Results

* **TEST_LONG_DELETION (OPEN):** Tested with the attached PR build (17.0.268-beta-test-10728) on iPhone 13 Mobile device and here is my observation: 1. Enabled Keyman as the system keyboard. 2. Opened Safari browser. Switched to the default iOS system keyboard. 3. Copied the text "Testing testing" from another PR's page on the test device. 4. Opened the Notes app and created a new note. Pasted the text copied from the PR's page. 5. Switched to Keyman Keyboard. 6. I confirmed that all the texts were deleted after pressing and holding down the Backspace key.

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_SHIFTED_CONTEXT_MULTITAP_IOS (PASSED):** Tested with the attached PR build (17.0.268-beta-test-10728) on iPhone 13 Mobile device and here is my observation: 1. Downloaded and installed diacritic_rota.kmp file in the Keyman In-App. 2. Switched to English - Diacritic 10-key Rota keyboard. 3. Followed the steps from 3 - 9 and verified that it showed the expected final result aềa. 4. Double-tapped the ENTER key. 5. Repeated Steps 3 - 9 for two times. Verified that the results for the first and last lines were identical with no extra lines produced after step 4.

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_SHIFTED_CONTEXT_MULTITAP_ANDROID (FAILED):** Tested with the attached PR build (17.0.268-beta-test-10728) on iPhone 13 Mobile device and here is my observation: 1. Downloaded and installed diacritic_rota.kmp file in the Keyman In-App. 2. Switched to English - Diacritic 10-key Rota keyboard. 3. Followed the steps from 3 - 9 and verified that it showed the expected final result aềa. 4. Double-tapped the ENTER key. 5. Repeated Steps 3 - 9 for the second time. Noticed that the Double-tap  the 8 key showed the number 8 on the text screen. Seems to be an issue.

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_NEW_PARAGRAPH_PREDICTIONS (FAILED):** Tested with the attached PR build (17.0.268-beta-test-10728) on an iPhone 13 Mobile device and here are my observation: 1. Typed a few paragraphs and applied suggestions while typing words. 2. I noticed that tapping on the word "many" in the banner incorrectly outputs the word "managed" on the text screen. 3. Additionally, long-pressing the period key did not bring up the subkey menu. 4. Sometimes, a key on the keyboard would get stuck and appear in blue color.

**Comment by jahorton** [comment]
> # Test Results
> 
> * **TEST_SHIFTED_CONTEXT_MULTITAP_ANDROID (FAILED):** Tested with the attached PR build (17.0.268-beta-test-10728) on iPhone 13 Mobile device and here is my observation: 1. Downloaded and installed diacritic_rota.kmp file in the Keyman In-App. 2. Switched to English - Diacritic 10-key Rota keyboard. 3. Followed the steps from 3 - 9 and verified that it showed the expected final result aềa. 4. Double-tapped the ENTER key. 5. Repeated Steps 3 - 9 for the second time. Noticed that the Double-tap  the 8 key showed the number 8 on the text screen. Seems to be an issue.
>        

> on iPhone 13 Mobile 

@bharanidharanj 

I'm pretty sure that's not an iPhone 13 in the screenshot.  Alas, I'm having trouble getting a local repro of this at the moment, so I actually might need details for the device you were using in order to reproduce the issue properly.

Alternatively, it might help to have a video of the repro - largely to confirm that both intended taps did land.  If only one did... yeah, you'd be stuck with that 8.

**Comment by jahorton** [comment]
> # Test Results
> 
>     * **TEST_NEW_PARAGRAPH_PREDICTIONS (FAILED):** Tested with the attached PR build (17.0.268-beta-test-10728) on an iPhone 13 Mobile device and here are my observation: 1. Typed a few paragraphs and applied suggestions while typing words. 2. I noticed that tapping on the word "many" in the banner incorrectly outputs the word "managed" on the text screen. 3. Additionally, long-pressing the period key did not bring up the subkey menu. 4. Sometimes, a key on the keyboard would get stuck and appear in blue color.
>        

@bharanidharanj I'm going to have to ask for video on this one.  I'm not currently able to reproduce it, and the applied suggestion is highlighted in your screenshot above.  There's an admittedly small chance of this, but the possibility exists that I'd need to replicate what you did _quite_ precisely to get the same effects, and video would be, by far, the most convenient way to provide that.

As for the other bits (sticky highlighting, period subkey menu), I've gone ahead and merged in recent changes that may be related, so... on that basis...

@keymanapp-test-bot retest TEST_SHIFTED_CONTEXT_MULTITAP_ANDROID TEST_NEW_PARAGRAPH_PREDICTIONS 

See my previous comment re: the Android-oriented redo request.

**Comment by bharanidharanj** [comment]
> > # Test Results
> > 
> > * **TEST_SHIFTED_CONTEXT_MULTITAP_ANDROID (FAILED):** Tested with the attached PR build (17.0.268-beta-test-10728) on iPhone 13 Mobile device and here is my observation: 1. Downloaded and installed diacritic_rota.kmp file in the Keyman In-App. 2. Switched to English - Diacritic 10-key Rota keyboard. 3. Followed the steps from 3 - 9 and verified that it showed the expected final result aềa. 4. Double-tapped the ENTER key. 5. Repeated Steps 3 - 9 for the second time. Noticed that the Double-tap  the 8 key showed the number 8 on the text screen. Seems to be an issue.
> >   
> 
> > on iPhone 13 Mobile
> 
> @bharanidharanj
> 
> I'm pretty sure that's not an iPhone 13 in the screenshot. Alas, I'm having trouble getting a local repro of this at the moment, so I actually might need details for the device you were using in order to reproduce the issue properly.
> 
> Alternatively, it might help to have a video of the repro - largely to confirm that both intended taps did land. If only one did... yeah, you'd be stuck with that 8.

@jahorton Oops! Sorry for the confusion. Yes. I have tested this in my Android Mobile (ver 13) device. Sure. I will add a video file for reference. Thanks.

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_SHIFTED_CONTEXT_MULTITAP_ANDROID (PASSED):** Tested with the attached PR build (17.0.268-beta-test-10728) in the Android Mobile device (ver 13) and here is my observation: 1. Downloaded and installed diacritic_rota.kmp file in the Keyman In-App. 2. Switched to English - Diacritic 10-key Rota keyboard. 3. Followed the steps from 3 - 9 and verified that it showed the expected final result aềa. 4. Double-tapped the ENTER key. 5. Repeated Steps 3 - 9 for the second time.  Verified that the results for the first and last lines were identical with no extra lines produced after step 4. However, there was a delay or slowness experienced (so that the number appeared on the screen) when double-tapping the 8 or 9 key on the keyboard. I have attached the Video file for reference. Thanks. 

https://github.com/keymanapp/keyman/assets/19683143/b1a7af51-f956-4c7d-8da1-1c9d42a6c38c

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_NEW_PARAGRAPH_PREDICTIONS (PASSED):** Retested with the attached PR build (17.0.268-beta-test-10728) on an iPhone 13 Mobile device and here are my observations:  Tried a lot to replicate the issue that I have already posted, but this time I did not see any predictive text stuck during the process. Also, I tried to save this in a video file (3 minute .mp4 video) but its size more than 60MB. I tried to compress the video file but it only compressed upto 30MB. I was not able to reproduce this issue. Anyhow, I will keep an eye on this issue and will post as a separate one if it is reproducible.

File: ios/engine/KMEI/KeymanEngine/resources/Keyman.bundle/Contents/Resources/ios-host.js [hunk 1]
```diff
@@ -276,33 +276,21 @@ function doResetContext() {
...
+    var end = pos + length;
+    // undefined + <number> => NaN.
+    end = isNaN(end) ? undefined : end;
+
+    const shouldReset = keyman.context.updateContext(text, start, end);
+    if(!doSync || shouldReset) {
         keyman.resetContext();
     }
-
-    return keyman.context.getText();
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngine/resources/Keyman.bundle/Contents/Resources/ios-host.js:305]: Was this never used?

File: ios/engine/KMEI/KeymanEngine/resources/Keyman.bundle/Contents/Resources/ios-host.js [hunk 2]
```diff
@@ -276,33 +276,21 @@ function doResetContext() {
...
-
+function setKeymanContext(text, doSync, pos, length) {
+    // console.log(`setKeymanContext(${JSON.stringify(text)}, ${doSync}, ${pos}, ${length})`);
     if(text == undefined) {
         text = '';
     }
 
-    if(keyman.context.getText() != text) {
-        keyman.context.setText(text);
+    var start = pos;
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngine/resources/Keyman.bundle/Contents/Resources/ios-host.js:285]:
```suggestion
```

Just renaming `pos` to `start` and eliminating this unnecessary variable.

File: ios/engine/KMEI/KeymanEngine/resources/Keyman.bundle/Contents/Resources/ios-host.js [hunk 3]
```diff
@@ -276,33 +276,21 @@ function doResetContext() {
...
-      keyman.context.setSelection(start, end);
-      keyman.resetContext();
-    }
-}
-
-function setKeymanVal(text) {
-    //console.log('setKeymanVal('+JSON.stringify(text)+')');
-
+function setKeymanContext(text, doSync, pos, length) {
+    // console.log(`setKeymanContext(${JSON.stringify(text)}, ${doSync}, ${pos}, ${length})`);
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngine/resources/Keyman.bundle/Contents/Resources/ios-host.js:280]:
```suggestion
function setKeymanContext(text, doSync, start, length) {
    // console.log(`setKeymanContext(${JSON.stringify(text)}, ${doSync}, ${start}, ${length})`);
```

**Review by mcdurdin — APPROVED** [review]
I think LGTM. I don't really understand all the tendrils here -- there's a whole lot of complexity both in the original code and in the changes, and I don't think I have a full handle on all the interactions.

> Upon request, I suppose I could create a highly-focused test keyboard for validating such a case. That said, it does seem pretty niche - I don't think we currently have a keyboard with such rules. (As a result, I haven't prioritized testing this scenario.) Part of me wonders if the best course of action would be to spin this concern off into a separate issue so that it doesn't hold this PR up from fixing a lot of cases that are far more commonly seen.

Yes, sounds sensible, probably turn it into an issue.

File: ios/engine/KMEI/KeymanEngine/resources/Keyman.bundle/Contents/Resources/ios-host.js [hunk 4]
```diff
@@ -276,33 +276,21 @@ function doResetContext() {
...
     if(text == undefined) {
         text = '';
     }
 
-    if(keyman.context.getText() != text) {
-        keyman.context.setText(text);
+    var start = pos;
+    var end = pos + length;
+    // undefined + <number> => NaN.
+    end = isNaN(end) ? undefined : end;
```
[the remark this answers is not in the document]
Comment (jahorton) [reply ios/engine/KMEI/KeymanEngine/resources/Keyman.bundle/Contents/Resources/ios-host.js:288]: Will fix the `var` bit.

The comment isn't explicit about it, but both `pos` and `length` can be `undefined`.  The same property holds if both are undefined at the same time.

[the remark this answers is not in the document]
Comment (mcdurdin) [reply ios/engine/KMEI/KeymanEngine/resources/Keyman.bundle/Contents/Resources/ios-host.js:288]: OK, that's pretty ugly and side-effecty. Let's make sure we add a comment to that effect.

File: ios/engine/KMEI/KeymanEngine/resources/Keyman.bundle/Contents/Resources/ios-host.js (hunk 1 above)
Comment (jahorton) [reply ios/engine/KMEI/KeymanEngine/resources/Keyman.bundle/Contents/Resources/ios-host.js:305]: Yes.  Here's the only thing calling it:

https://github.com/keymanapp/keyman/blob/66142be3930c24f996939f7496a377f23223e3c3/ios/engine/KMEI/KeymanEngine/Classes/Keyboard/KeymanWebViewController.swift#L249

Note how nothing is checking for a result.  To be extra clear here, that would be done by the second argument if one were provided.

https://developer.apple.com/documentation/webkit/wkwebview/1415017-evaluatejavascript

No callback means no need for a return value.

**Comment by jahorton** [comment]
> > Upon request, I suppose I could create a highly-focused test keyboard for validating such a case. That said, it does seem pretty niche - I don't think we currently have a keyboard with such rules. (As a result, I haven't prioritized testing this scenario.) Part of me wonders if the best course of action would be to spin this concern off into a separate issue so that it doesn't hold this PR up from fixing a lot of cases that are far more commonly seen.
> 
> Yes, sounds sensible, probably turn it into an issue.

I've made a basic issue for this at #10761.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.271-beta](https://keyman.com/downloads/releases/beta/17.0.271)

**Pull request merged**